src/main/scala/sigmastate/lang/SigmaTyper.scala

```diff
@@ -339,17 +340,11 @@ class SigmaTyper {
           (newNode: (Value[T], Value[T]) => SValue): SValue = {
     val l1 = assignType(env, l).asValue[T]
     val r1 = assignType(env, r).asValue[T]
-    (l1.tpe, r1.tpe) match {
```

Comment:
@ergomorphic check out my take on reusing type constraining logic from `SigmaBuilded` in `SigmaTyper`.